extendedcopy.go

```diff
@@ -0,0 +1,123 @@
+limitations under the License.
+*/
+
+package oras
+
+import (
+	"context"
+	"errors"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/internal/descriptor"
+)
+
+// ExtendedCopy copies the directed acyclic graph (DAG) that are reachable from
+// the given tagged node from the source TraceableTarget to the destination Target.
+// The destination reference will be the same as the source reference if the
+// destination reference is left blank.
+// Returns the descriptor of the tagged node on successful copy.
+func ExtendedCopy(ctx context.Context, src TraceableTarget, srcRef string, dst Target, dstRef string) (ocispec.Descriptor, error) {
+	if src == nil {
+		return ocispec.Descriptor{}, errors.New("nil source traceable target")
+	}
+	if dst == nil {
+		return ocispec.Descriptor{}, errors.New("nil destination target")
+	}
+	if dstRef == "" {
+		dstRef = srcRef
+	}
+
+	node, err := src.Resolve(ctx, srcRef)
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+
+	if err := ExtendedCopyGraph(ctx, src, dst, node); err != nil {
+		return ocispec.Descriptor{}, err
+	}
+
+	if err := dst.Tag(ctx, node, dstRef); err != nil {
+		return ocispec.Descriptor{}, err
+	}
+
+	return node, nil
+}
+
+// ExtendedCopyGraph copies the directed acyclic graph (DAG) that are reachable
+// from the given node from the source traceable-CAS to the destination CAS.
+func ExtendedCopyGraph(ctx context.Context, src content.TraceableStorage, dst content.Storage, node ocispec.Descriptor) error {
+	exists, err := src.Exists(ctx, node)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		return errdef.ErrNotFound
+	}
+
+	// find the root nodes traceable from the current node
+	roots := make(map[descriptor.Descriptor]ocispec.Descriptor)
+	if err := traceRoots(ctx, src, roots, node); err != nil {
+		return err
+	}
+
+	// copy the sub-DAGs rooted by the root nodes
+	for _, root := range roots {
+		if err := CopyGraph(ctx, src, dst, root); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// traceRoots traces the root nodes from the given node,
+// and records them in the given map.
+func traceRoots(ctx context.Context, storage content.TraceableStorage, roots map[descriptor.Descriptor]ocispec.Descriptor, node ocispec.Descriptor) error {
+	exists, err := storage.Exists(ctx, node)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		return errdef.ErrNotFound
+	}
+
+	upEdges, err := storage.UpEdges(ctx, node)
+	if err != nil {
+		return err
+	}
+
+	// The current node has no parent node,
+	// which means it is a root node of a sub-DAG.
+	if len(upEdges) == 0 {
+		key := descriptor.FromOCI(node)
+		if _, exists := roots[key]; !exists {
+			roots[key] = node
+		}
+		return nil
+	}
+
+	// The current node has parents nodes. Keep tracing from the parent nodes.
+	for _, upEdge := range upEdges {
+		if err := traceRoots(ctx, storage, roots, upEdge); err != nil {
```

Comment:
It is a recursive implementation. Could we make it iterative? Meanwhile, can we optimize the algorithm to address something like

```
  A   H
 / \ /
v   v
B   C
 \ /
  v
  D
 / \
v   v
E   F
 \ /
  v
  G
```
Given `G`, the algorithm should output `A` and `H` while other nodes should be only traversed once?

Reply:
Fixed.